### PR TITLE
enhancement/integrate-store-chat

### LINKED ISSUE #196: Integrate store protocol into chat2 example
# Problem

As a user of chat2 example app I can't get historical messages.

# Acceptance criteria

In examples/chat2.nim, act as a storeclient and pull down messages from store server node. Show these messages in the beginning of the chat.

### PR DESCRIPTION
**Pull request by decanus**
closes #196

### CODE REVIEW

**Review by oskarth — COMMENTED** [review]
Looks reasonable, does it work?

1. Need to update instructions on Dingpu docs too for running nodes
2. How does it deal with race conditions? E.g. slow response from store node, does it show historical message after other messages?

**Comment by oskarth** [comment]
^ above

Also what is blocking this from no longer being a draft?

**Comment by decanus** [comment]
@oskarth sorry I thought I had responded. For number 2 I am not yet fully sure how it deals with it. I believe it would show the historical messages after the filtered one if they are timed like that.

**Review by oskarth — CHANGES_REQUESTED** [review]
Point 1 from 4 days ago not resolved. Please address feedback before asking for re-reviews.

**Comment by oskarth** [comment]
> . For number 2 I am not yet fully sure how it deals with it

Seems like a problem. I thought we wanted to address this by moving to REQ-RES, but doesn't seem like we have improved the situation?

**Comment by decanus** [comment]
> > . For number 2 I am not yet fully sure how it deals with it
> 
> 
> 
> Seems like a problem. I thought we wanted to address this by moving to REQ-RES, but doesn't seem like we have improved the situation?

It is REQ-RES...

**Comment by oskarth** [comment]
If it is REQ RES then why are we not sure how it orders messages for what should be blocking operations?

**Comment by decanus** [comment]
@oskarth looking at the code now again, it does block so messages are ordered.

**Comment by oskarth** [comment]
Once instructions are updated, I'll review

**Comment by decanus** [comment]
@oskarth I added a quick note to the docs. But it doesn't actually require anything in running from how you did it.

File: docs/tutorial/dingpu.md [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 ## Basic chat usage
```
Comment (oskarth) [inline docs/tutorial/dingpu.md:3]: Why don't you have to add a command line flag to indicate to use that node as store/filter node? Doesn't make sense to me

**Review by oskarth — APPROVED** [review]

**Comment by oskarth** [comment]
Discussed in call. The store protocol is inferred, by connecting to it directly interactively we don't need to supply staticnode or storenode argument. @decanus to test it but appears to be good enough.

(Alternative would be to be explicit about `staticnode` and/or `storenode`, then leave the interactive option open for e.g. local nodes)

**Comment by decanus** [comment]
@oskarth so I was testing and I seem to be having some issue, need to investigate further. have you encountered:

```
Error: unhandled exception: index out of bounds, the container is empty [IndexError]
```

while trying to run the app?

File: examples/v2/chat2.nim [hunk 2]
```diff
@@ -55,15 +55,16 @@ proc initAddress(T: type MultiAddress, str: string): T =
...
-  let remotePeer = PeerInfo.init(parts[^1], [multiAddr])
+  result = PeerInfo.init(parts[^1], [multiAddr])
 
-  echo &"dialing peer: {multiAddr}"
+# NOTE Dialing on WakuRelay specifically
+proc dialPeer(c: Chat, peer: PeerInfo) {.async.} =
+  echo &"dialing peer: {peer.peerId}"
   # XXX Discarding conn, do we want to keep this here?
-  discard await c.node.switch.dial(remotePeer, WakuRelayCodec)
+  discard await c.node.switch.dial(peer, WakuRelayCodec)
```
Comment (oskarth) [inline examples/v2/chat2.nim:67]: So this is only calling on WakuRelayCodec, you might need something else for Store?

File: examples/v2/chat2.nim [hunk 3]
```diff
@@ -162,10 +165,22 @@ proc processInput(rfd: AsyncFD, rng: ref BrHmacDrbgContext) {.async.} =
   let listenStr = $peerInfo.addrs[0] & "/p2p/" & $peerInfo.peerId
   echo &"Listening on\n {listenStr}"
 
+  let topic = cast[Topic](DefaultContentTopic)
+  let multiAddr = MultiAddress.initAddress(conf.storenode)
+  let parts = conf.storenode.split("/")
+
+  node.wakuStore.setPeer(PeerInfo.init(parts[^1], [multiAddr]))
```
Comment (oskarth) [inline examples/v2/chat2.nim:172]: Maybe this is where the `Error: unhandled exception: index out of bounds, the container is empty [IndexError]` comes from? If storenode isn't specified correctly

**Comment by oskarth** [comment]
> @oskarth so I was testing and I seem to be having some issue, need to investigate further. have you encountered:
> 
> ```
> Error: unhandled exception: index out of bounds, the container is empty [IndexError]
> ```
> 
> while trying to run the app?

Not on master no, but that's a fairly standard error. Probably from a seq not being populated somewhere/off by one in your branch?

**Comment by decanus** [comment]
blocked by https://github.com/status-im/nim-waku/issues/221

**Comment by oskarth** [comment]
I don't see how it is blocked by that issue?

File: examples/v2/chat2.nim (hunk 2 above)
Comment (decanus) [reply examples/v2/chat2.nim:67]: no, because the WakuStore protocol dials by itself.

**Comment by decanus** [comment]
@oskarth ok this is ready to go. There was a bug in the node causing it not to work. I've resolved that and will deploy new cluster on merge.

**Comment by oskarth** [comment]
OK. What was the bug on the node? (wakunode2?)

**Comment by oskarth** [comment]
Yeah this doesn't work. It breaks chat2 with previous behaviour.

 With this commit in master I get this error:

```
DBG 2020-10-15 13:00:18.122+08:00 Started libp2p node                        topics="switch" tid=16808 peer="(peerId: \"16Uiu2HAm2hMhH15gFRzMsemym58ZqPJkmMBVS29jd2HZenb4wXJb\", addrs: @[\"/ip4/0.
0.0.0/tcp/60000\"], protocols: @[], protoVersion: \"\", agentVersion: \"\")"                                                                                                                       
DBG 2020-10-15 13:00:18.122+08:00 subscribe                                  topics="wakurelay" tid=16808 pubSubTopic=waku                                                                         
INF 2020-10-15 13:00:18.122+08:00 rebalance mesh from subscribe              topics="gossipsub" tid=16808                                                                                          
INF 2020-10-15 13:00:18.122+08:00 rebalancing mesh                           topics="gossipsub" tid=16808 mesh=0 gossipsub=0 topic=waku                                                            
INF 2020-10-15 13:00:18.122+08:00 Mesh npeers                                topics="gossipsub" tid=16808 npeers=0 topic=waku mesh=0 gossipsub=0                                                   
INF 2020-10-15 13:00:18.122+08:00 PeerInfo                                   topics="wakunode" tid=16808 peerId=16*b4wXJb addrs=@[/ip4/0.0.0.0/tcp/60000]                                          
INF 2020-10-15 13:00:18.122+08:00 Listening on                               topics="wakunode" tid=16808 full=/ip4/0.0.0.0/tcp/60000/p2p/16Uiu2HAm2hMhH15gFRzMsemym58ZqPJkmMBVS29jd2HZenb4wXJb     
Listening on                                                                                                                                                                                       
 /ip4/0.0.0.0/tcp/60000/p2p/16Uiu2HAm2hMhH15gFRzMsemym58ZqPJkmMBVS29jd2HZenb4wXJb                                                                                                                  
Traceback (most recent call last, using override)                                                                                                                                                  /home/oskarth/git/status-im/nim-waku/vendor/nim-eth/eth/common/eth_types.nim(577) main           
/home/oskarth/git/status-im/nim-waku/vendor/nim-eth/eth/common/eth_types.nim(570) NimMain        
/home/oskarth/git/status-im/nim-waku/vendor/nimbus-build-system/vendor/Nim/lib/system/excpt.nim(407) reportUnhandledError                                                                          
/home/oskarth/git/status-im/nim-waku/vendor/nimbus-build-system/vendor/Nim/lib/system/excpt.nim(358) reportUnhandledErrorAux                                                                       
Error: unhandled exception: Invalid bootstrap node multi-address [ValueError]                                                                                                                      
DBG 2020-10-15 13:00:20.262+08:00 UPnP: deleted port mapping                 topics="nat" tid=16808 externalPort=60000 internalPort=60000 protocol=TCP                                             
DBG 2020-10-15 13:00:20.269+08:00 UPnP: deleted port mapping                 topics="nat" tid=16808 externalPort=60000 internalPort=60000 protocol=UDP  
```

Without it it works.

I don't want to debug this. My guess would be you only tested with storenode specified or something. Do again and do it right. Reverting from master.

**Pull request merged**